File: sdk/python/tempmail_sdk/providers/moakt.py
"""
moakt.com：GET 语言首页与收件箱 HTML；凭证为 tm_session 等 Cookie（序列化在 token 内）；
列表解析 /{locale}/email/{uuid}，正文 GET .../html 解析 .email-body。
"""

import base64
import html
import json
import re
import secrets
from typing import Dict, List, Optional, Tuple

import requests

from ..config import get_config
from .. import http
from ..normalize import normalize_email
from ..types import Email, EmailInfo

CHANNEL = "moakt"
ORIGIN = "https://www.moakt.com"
TOK_PREFIX = "mok1:"

_EMAIL_DIV_RE = re.compile(r'(?is)<div\s+id="email-address"\s*>([^<]+)</div>')
_DOMAIN_OPTION_RE = re.compile(r'(?is)<option\s+value="([^"]+)">\s*@[^<]+</option>')
_MAIL_DOMAIN_RE = re.compile(
    r"(?i)^[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?(?:\.[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?)+$"
)
_HREF_EMAIL_RE = re.compile(
    r'href="(/[^"]+/email/[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12})"'
)
_TITLE_RE = re.compile(r'(?is)<li\s+class="title"\s*>([^<]*)</li>')
_DATE_RE = re.compile(r'(?is)<li\s+class="date"[^>]*>[\s\S]*?<span[^>]*>([^<]+)</span>')
_SENDER_RE = re.compile(
    r'(?is)<li\s+class="sender"[^>]*>[\s\S]*?<span[^>]*>([\s\S]*?)</span>\s*</li>'
)
_BODY_OPEN_RE = re.compile(r'(?is)<div\s+class="email-body"\s*>')


def _extract_body_html(page: str) -> str:
    """使用栈式深度匹配提取 email-body div 的完整内部 HTML，
    避免非贪婪正则在嵌套 div 时截断正文。
    """
    m = _BODY_OPEN_RE.search(page)
    if not m:
        return ""
    start = m.end()
    pos = start
    depth = 1
    while pos < len(page) and depth > 0:
        next_open = page.find("<div", pos)
        next_close = page.find("</div>", pos)
        if next_close == -1:
            break
        if next_open != -1 and next_open < next_close:
            depth += 1
            pos = next_open + 4
        else:
            depth -= 1
            if depth == 0:
                return page[start:next_close].strip()
            pos = next_close + 6
    return ""
_FROM_ADDR_RE = re.compile(r"<([a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,})>")
_TAG_RE = re.compile(r"<[^>]+>")

_DEFAULT_HEADERS = {
    "Accept-Language": "zh-CN,zh;q=0.9,en;q=0.8,en-GB;q=0.7,en-US;q=0.6",
    "Cache-Control": "no-cache",
    "DNT": "1",
    "Pragma": "no-cache",
    "Upgrade-Insecure-Requests": "1",
}


def _request_parts(domain: Optional[str]) -> Tuple[str, str]:
    s = (domain or "").strip()
    if not s or any(c in s for c in "/?#\\"):
        return "zh", ""
    if _MAIL_DOMAIN_RE.match(s):
        return "zh", s.lstrip("@").lower()
    return s, ""


def _parse_server_domains(page: str) -> set[str]:
    return {
        m.group(1).strip().lstrip("@").lower()
        for m in _DOMAIN_OPTION_RE.finditer(page)
        if m.group(1).strip()
    }


def _random_local(length: int = 12) -> str:
    chars = "abcdefghijklmnopqrstuvwxyz0123456789"
    return "".join(secrets.choice(chars) for _ in range(length))


def _email_domain(email: str) -> str:
    _, sep, domain = email.rpartition("@")
    return domain.strip().lower() if sep else ""


def _bare_get(url: str, headers: dict, **kwargs) -> requests.Response:
    """使用共享HTTP客户端发起GET请求（复用连接池）"""
    return http.get(url, headers=headers, **kwargs)


def _bare_post(url: str, headers: dict, **kwargs) -> requests.Response:
    """使用共享HTTP客户端发起POST请求（复用连接池）"""
    return http.post(url, headers=headers, **kwargs)


def _parse_cookie_map(hdr: str) -> Dict[str, str]:
    m: Dict[str, str] = {}
    for part in hdr.split(";"):
        part = part.strip()
        if not part or "=" not in part:
            continue
        k, _, v = part.partition("=")
        k, v = k.strip(), v.strip()
        if k:
            m[k] = v
    return m


def _merge_cookie_hdr(prev: str, resp: requests.Response) -> str:
    d = _parse_cookie_map(prev)
    d.update(resp.cookies.get_dict())
    return "; ".join(f"{k}={d[k]}" for k in sorted(d.keys()))


def _page_headers(referer: str, ua: str) -> dict:
    return {
        **_DEFAULT_HEADERS,
        "User-Agent": ua,
        "Accept": "text/html,application/xhtml+xml,application/xml;q=0.9,image/avif,image/webp,image/apng,*/*;q=0.8",
        "Referer": referer,
    }


def _encode_sess(locale: str, cookie_hdr: str) -> str:
    raw = json.dumps({"l": locale, "c": cookie_hdr}, separators=(",", ":")).encode(
        "utf-8"
    )
    return TOK_PREFIX + base64.standard_b64encode(raw).decode("ascii")


def _decode_sess(tok: str) -> Tuple[str, str]:
    if not tok.startswith(TOK_PREFIX):
        raise ValueError("moakt: invalid session token")
    try:
        data = base64.standard_b64decode(tok[len(TOK_PREFIX) :].encode("ascii"))
        o = json.loads(data.decode("utf-8"))
    except (json.JSONDecodeError, ValueError) as e:
        raise ValueError("moakt: invalid session token") from e
    loc = (o.get("l") or "").strip()
    c = (o.get("c") or "").strip()
    if not loc or not c:
        raise ValueError("moakt: invalid session token")
    return loc, c


def _parse_inbox_email(html_s: str) -> str:
    m = _EMAIL_DIV_RE.search(html_s)
    if not m:
        raise RuntimeError("moakt: email-address not found")
    addr = html.unescape(m.group(1).strip())
    if not addr:
        raise RuntimeError("moakt: empty email-address")
    return addr


def _strip_tags(s: str) -> str:
    return _TAG_RE.sub(" ", s).strip()


def _list_mail_ids(html_s: str) -> List[str]:
    seen = set()
    out: List[str] = []
    for m in _HREF_EMAIL_RE.finditer(html_s):
        path = m.group(1)
        if "/delete" in path:
            continue
        mid = path.rsplit("/", 1)[-1]
        if len(mid) == 36 and mid not in seen:
            seen.add(mid)
            out.append(mid)
    return out


def _parse_detail(page: str, mid: str, recipient: str) -> dict:
    from_s = ""
    sm = _SENDER_RE.search(page)
    if sm:
        inner = html.unescape(sm.group(1))
        from_s = _strip_tags(inner)
        em = _FROM_ADDR_RE.search(inner)
        if em:
            from_s = em.group(1).strip()
    subj = ""
    tm = _TITLE_RE.search(page)
    if tm:
        subj = html.unescape(tm.group(1).strip())
    date_s = ""
    dm = _DATE_RE.search(page)
    if dm:
        date_s = html.unescape(dm.group(1).strip())
    body = _extract_body_html(page)
    return {
        "id": mid,
        "to": recipient,
        "from": from_s,
        "subject": subj,
        "date": date_s,
        "html": body,
    }


def generate_email(domain: Optional[str] = None, **kwargs) -> EmailInfo:
    loc, mail_domain = _request_parts(domain)
    base = f"{ORIGIN}/{loc}"
    inbox = f"{base}/inbox"
    c = get_config()
    ua = (c.headers or {}).get("User-Agent") or (
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) "
        "Chrome/146.0.0.0 Safari/537.36 Edg/146.0.0.0"
    )

    r1 = _bare_get(base, _page_headers(base, ua))
    r1.raise_for_status()
    cookie_hdr = _merge_cookie_hdr("", r1)

    if mail_domain:
        if mail_domain not in _parse_server_domains(r1.text):
            raise RuntimeError(f"moakt: unsupported domain {mail_domain}")
        post_data: dict = {
            "setemail": "",
            "username": _random_local(),
            "domain": mail_domain,
            "preferred_domain": "",
        }
    else:
        post_data = {"random": "1"}

    # POST /inbox 创建邮箱，捕获 302 中的 tm_session cookie
    r2 = _bare_post(
        inbox,
        {
            **_page_headers(base, ua),
            "Content-Type": "application/x-www-form-urlencoded",
            "Cookie": cookie_hdr,
        },
        data=post_data,
        allow_redirects=False,
    )
    cookie_hdr = _merge_cookie_hdr(cookie_hdr, r2)

    if "tm_session" not in _parse_cookie_map(cookie_hdr):
        raise RuntimeError("moakt: missing tm_session cookie")

    # GET /inbox 获取邮箱地址
    r3 = _bare_get(
        inbox,
        {**_page_headers(base, ua), "Cookie": cookie_hdr},
    )
    r3.raise_for_status()
    cookie_hdr = _merge_cookie_hdr(cookie_hdr, r3)
    html_s = r3.text

    email = _parse_inbox_email(html_s)
    if mail_domain and _email_domain(email) != mail_domain:
        raise RuntimeError(
            f"moakt: domain mismatch expected={mail_domain} actual={_email_domain(email)}"
        )
    tok = _encode_sess(loc, cookie_hdr)
    return EmailInfo(channel=CHANNEL, email=email, _token=tok)


def get_emails(email: str, token: str, **kwargs) -> List[Email]:
    loc, cookie_hdr = _decode_sess(token)
    inbox = f"{ORIGIN}/{loc}/inbox"
    c = get_config()
    ua = (c.headers or {}).get("User-Agent") or (
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) "
        "Chrome/146.0.0.0 Safari/537.36 Edg/146.0.0.0"
    )
    base_ref = f"{ORIGIN}/{loc}"

    r = _bare_get(
        inbox,
        {**_page_headers(base_ref, ua), "Cookie": cookie_hdr},
    )
    r.raise_for_status()
    ids = _list_mail_ids(r.text)
    out: List[Email] = []
    for mid in ids:
        detail = f"{ORIGIN}/{loc}/email/{mid}/html"
        refer = f"{ORIGIN}/{loc}/email/{mid}"
        rd = _bare_get(
            detail,
            {**_page_headers(refer, ua), "Cookie": cookie_hdr},
        )
        if rd.status_code != 200:
            continue
        raw = _parse_detail(rd.text, mid, email)
        out.append(normalize_email(raw, email))
    return out

File: sdk/ruby/lib/tempmail_sdk/providers/moakt.rb
# frozen_string_literal: true

require "base64"
require "cgi"
require "json"
require "uri"

module TempmailSdk
  module Providers
    # moakt.com 渠道实现（HTML 抓取型）
    #
    # 流程：
    #   1. GET 语言首页，收集初始 Cookie 与可用域名 <option> 列表；
    #   2. POST /{locale}/inbox 创建邮箱（随机域名或指定域名），从 302 响应捕获 tm_session Cookie；
    #   3. GET /{locale}/inbox 解析 #email-address 得到邮箱地址；
    #   4. 会话凭证（locale + 序列化 Cookie 头）经 base64 打包进 token；
    #   5. 收信时列表解析 /{locale}/email/{uuid}，正文 GET .../html 解析 .email-body。
    #
    # 多个渠道别名（moakt.cc/moakt.co/moakt.ws/tmail.ws 等）共用此实现，仅固定域名不同。
    module Moakt
      CHANNEL = "moakt"
      ORIGIN = "https://www.moakt.com"
      TOK_PREFIX = "mok1:"

      # 默认 User-Agent（config.headers 未指定时使用）
      DEFAULT_UA = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 " \
                   "(KHTML, like Gecko) Chrome/146.0.0.0 Safari/537.36 Edg/146.0.0.0"

      # 页面请求的固定基础头
      DEFAULT_HEADERS = {
        "Accept-Language" => "zh-CN,zh;q=0.9,en;q=0.8,en-GB;q=0.7,en-US;q=0.6",
        "Cache-Control" => "no-cache",
        "DNT" => "1",
        "Pragma" => "no-cache",
        "Upgrade-Insecure-Requests" => "1"
      }.freeze

      EMAIL_DIV_RE = /<div\s+id="email-address"\s*>([^<]+)<\/div>/im
      DOMAIN_OPTION_RE = /<option\s+value="([^"]+)">\s*@[^<]+<\/option>/im
      MAIL_DOMAIN_RE = /\A[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?(?:\.[a-z0-9](?:[a-z0-9-]{0,61}[a-z0-9])?)+\z/i
      HREF_EMAIL_RE = /href="(\/[^"]+\/email\/[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12})"/
      TITLE_RE = /<li\s+class="title"\s*>([^<]*)<\/li>/im
      DATE_RE = /<li\s+class="date"[^>]*>[\s\S]*?<span[^>]*>([^<]+)<\/span>/im
      SENDER_RE = /<li\s+class="sender"[^>]*>[\s\S]*?<span[^>]*>([\s\S]*?)<\/span>\s*<\/li>/im
      BODY_RE = /<div\s+class="email-body"\s*>([\s\S]*?)<\/div>/im
      FROM_ADDR_RE = /<([a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,})>/
      TAG_RE = /<[^>]+>/

      module_function

      # 解析域名参数，返回 [locale, mail_domain]
      # - 空或含非法字符：语言页取默认 "zh"，随机域名
      # - 形如域名：locale="zh"，使用指定域名
      # - 其它字符串：视为 locale，随机域名
      def request_parts(domain)
        s = domain.to_s.strip
        return ["zh", ""] if s.empty? || s.match?(/[\/?#\\]/)
        return ["zh", s.sub(/\A@/, "").downcase] if s.match?(MAIL_DOMAIN_RE)

        [s, ""]
      end

      # 从首页 HTML 收集服务端提供的可用域名集合
      def parse_server_domains(page)
        page.scan(DOMAIN_OPTION_RE)
            .map { |m| m[0].strip.sub(/\A@/, "").downcase }
            .reject(&:empty?)
      end

      # 生成随机本地部分（小写字母 + 数字）
      def random_local(length = 12)
        chars = ("a".."z").to_a + ("0".."9").to_a
        Array.new(length) { chars.sample }.join
      end

      # 提取邮箱地址的域名部分（小写）
      def email_domain(email)
        _head, sep, domain = email.to_s.rpartition("@")
        sep.empty? ? "" : domain.strip.downcase
      end

      # 将 Cookie 头字符串解析为 name => value 映射
      def parse_cookie_map(hdr)
        map = {}
        hdr.to_s.split(";").each do |part|
          part = part.strip
          next unless part.include?("=")

          k, v = part.split("=", 2)
          k = k.strip
          map[k] = v.to_s.strip unless k.empty?
        end
        map
      end

      # 从响应的 Set-Cookie 头提取本次新设置的 Cookie
      def cookies_from_response(resp)
        map = {}
        resp.set_cookies.each do |line|
          pair = line.split(";", 2).first.to_s.strip
          next unless pair.include?("=")

          k, v = pair.split("=", 2)
          k = k.strip
          map[k] = v.to_s.strip unless k.empty?
        end
        map
      end

      # 合并旧 Cookie 头与响应新设置的 Cookie，按 key 排序拼接（保证稳定）
      def merge_cookie_hdr(prev, resp)
        d = parse_cookie_map(prev)
        d.merge!(cookies_from_response(resp))
        d.keys.sort.map { |k| "#{k}=#{d[k]}" }.join("; ")
      end

      # 构造页面请求头
      def page_headers(referer, ua)
        DEFAULT_HEADERS.merge(
          "User-Agent" => ua,
          "Accept" => "text/html,application/xhtml+xml,application/xml;q=0.9," \
                      "image/avif,image/webp,image/apng,*/*;q=0.8",
          "Referer" => referer
        )
      end

      # 从 config 取 UA，缺省用内置默认值
      def user_agent
        (Config.get_config.headers || {})["User-Agent"] || DEFAULT_UA
      end

      # 将会话（locale + Cookie 头）编码为不透明 token
      def encode_sess(locale, cookie_hdr)
        raw = JSON.generate("l" => locale, "c" => cookie_hdr)
        TOK_PREFIX + Base64.strict_encode64(raw)
      end

      # 解码 token，返回 [locale, cookie_hdr]
      def decode_sess(tok)
        raise "moakt: invalid session token" unless tok.to_s.start_with?(TOK_PREFIX)

        data = Base64.strict_decode64(tok[TOK_PREFIX.length..])
        o = JSON.parse(data)
        loc = o["l"].to_s.strip
        c = o["c"].to_s.strip
        raise "moakt: invalid session token" if loc.empty? || c.empty?

        [loc, c]
      rescue ArgumentError, JSON::ParserError
        raise "moakt: invalid session token"
      end

      # 从收件箱页解析邮箱地址
      def parse_inbox_email(html_s)
        m = html_s.match(EMAIL_DIV_RE)
        raise "moakt: email-address not found" unless m

        addr = CGI.unescapeHTML(m[1].strip)
        raise "moakt: empty email-address" if addr.empty?

        addr
      end

      # 去除 HTML 标签，压缩为文本
      def strip_tags(s)
        s.gsub(TAG_RE, " ").strip
      end

      # 从收件箱页提取邮件 UUID 列表（去重、排除删除链接）
      def list_mail_ids(html_s)
        seen = {}
        out = []
        html_s.scan(HREF_EMAIL_RE).each do |cap|
          path = cap[0]
          next if path.include?("/delete")

          mid = path.split("/").last
          if mid.length == 36 && !seen[mid]
            seen[mid] = true
            out << mid
          end
        end
        out
      end

      # 解析邮件详情页，返回标准化前的原始 Hash
      def parse_detail(page, mid, recipient)
        from_s = ""
        if (sm = page.match(SENDER_RE))
          inner = CGI.unescapeHTML(sm[1])
          from_s = strip_tags(inner)
          if (em = inner.match(FROM_ADDR_RE))
            from_s = em[1].strip
          end
        end

        subj = ""
        subj = CGI.unescapeHTML(page.match(TITLE_RE)[1].strip) if page.match?(TITLE_RE)

        date_s = ""
        date_s = CGI.unescapeHTML(page.match(DATE_RE)[1].strip) if page.match?(DATE_RE)

        body = ""
        body = page.match(BODY_RE)[1].strip if page.match?(BODY_RE)

        {
          "id" => mid,
          "to" => recipient,
          "from" => from_s,
          "subject" => subj,
          "date" => date_s,
          "html" => body
        }
      end

      # 创建临时邮箱
      # @param domain [String, nil] 指定接入域名或语言标识
      # @return [EmailInfo]
      def generate_email(domain = nil)
        loc, mail_domain = request_parts(domain)
        base = "#{ORIGIN}/#{loc}"
        inbox = "#{base}/inbox"
        ua = user_agent

        r1 = Http.get(base, headers: page_headers(base, ua))
        r1.raise_for_status
        cookie_hdr = merge_cookie_hdr("", r1)

        if mail_domain.empty?
          post_data = { "random" => "1" }
        else
          unless parse_server_domains(r1.body).include?(mail_domain)
            raise "moakt: unsupported domain #{mail_domain}"
          end

          post_data = {
            "setemail" => "",
            "username" => random_local,
            "domain" => mail_domain,
            "preferred_domain" => ""
          }
        end

        # POST /inbox 创建邮箱，从 302 响应捕获 tm_session Cookie（Net::HTTP 默认不跟随重定向）
        r2 = Http.post(
          inbox,
          headers: page_headers(base, ua).merge(
            "Content-Type" => "application/x-www-form-urlencoded",
            "Cookie" => cookie_hdr
          ),
          body: URI.encode_www_form(post_data)
        )
        cookie_hdr = merge_cookie_hdr(cookie_hdr, r2)

        unless parse_cookie_map(cookie_hdr).key?("tm_session")
          raise "moakt: missing tm_session cookie"
        end

        # GET /inbox 获取邮箱地址
        r3 = Http.get(inbox, headers: page_headers(base, ua).merge("Cookie" => cookie_hdr))
        r3.raise_for_status
        cookie_hdr = merge_cookie_hdr(cookie_hdr, r3)

        email = parse_inbox_email(r3.body)
        if !mail_domain.empty? && email_domain(email) != mail_domain
          raise "moakt: domain mismatch expected=#{mail_domain} actual=#{email_domain(email)}"
        end

        EmailInfo.new(channel: CHANNEL, email: email, token: encode_sess(loc, cookie_hdr))
      end

      # 收取邮件列表
      # @param email [String]
      # @param token [String]
      # @return [Array<Email>]
      def get_emails(email, token)
        loc, cookie_hdr = decode_sess(token)
        inbox = "#{ORIGIN}/#{loc}/inbox"
        base_ref = "#{ORIGIN}/#{loc}"
        ua = user_agent

        r = Http.get(inbox, headers: page_headers(base_ref, ua).merge("Cookie" => cookie_hdr))
        r.raise_for_status

        list_mail_ids(r.body).filter_map do |mid|
          detail = "#{ORIGIN}/#{loc}/email/#{mid}/html"
          refer = "#{ORIGIN}/#{loc}/email/#{mid}"
          rd = Http.get(detail, headers: page_headers(refer, ua).merge("Cookie" => cookie_hdr))
          next unless rd.status_code == 200

          Normalize.normalize_email(parse_detail(rd.body, mid, email), email)
        end
      end
    end
  end
end
